### DISCUSSION TITLE
Finite Strain and in echelon dynamics test

### DISCUSSION THREAD

**Discussion by ghost** [idaholab/moose#28865] [Q&A Modules: Solid mechanics] [answered]
Hello, 

We have trouble working with dynamics in finite strains when using initial conditions on displacement. In fact we can see the correct displacement at the first increment while the strain and stress computations are wrong. This in echelon test works fine in small strain even when coupled with phase field fracture. I wonder if it's because of the strain initialization at the ComputeFiniteStrain ? Thanks for any help.

**Comment by selarem** [comment]
Any help on this issue ?

**Comment by ghost** [comment]
When using an initial condition like : 
```
[Variables]
  [./disp_x]
  [./InitialCondition]
      type = FunctionIC
      function = initial_cond_func_x
    [../]
  [../]
  [./disp_y]
  [./InitialCondition]
      type = FunctionIC
      function = initial_cond_func_y
    [../]
  [../]
[]

[Functions]
[./initial_cond_func_x]
    type = SolutionFunction
    solution = s0
    from_variable = 'disp_x'
[../]
 [./initial_cond_func_y]
    type = SolutionFunction
    solution = s0
    from_variable = 'disp_y'
[../]
[]

[UserObjects]
  [./s0]
    type = SolutionUserObject
    mesh = Finitehalf_elastic_960_out.e #half_elastic_960_out.e
    timestep = LATEST
    execute_on = initial
  [../]
[]
```
And with the Material block as:
```
[Materials]
 [./elasticity_tensor]
    type = ComputeIsotropicElasticityTensor
    youngs_modulus = 3.09e9
    poissons_ratio = 0.35
  [../]
[./elasticStrain]
    type = ComputeFiniteStrainElasticStress
  [../]
[]
```

We got correct displacements but stress and strain are null which is strange.

**Reply by ghost** [reply]
could this be due to the initialization in the 'ComputeFiniteStrain.C' or 'ComputeIncrementalStrainBase.C'

**Reply by GiudGiud** [reply]
Hello

You need to use a proper restart using Checkpoints to restart stateful material properties. 

You can't use solution user objects to restart just some variables. Unless you put in more objects from the ProjectedMaterialPropertyStorage (or some name like this) action.

**Reply by ghost** [reply]
Ok thanks! I did this using Checkpoint and restart_file_base but I got this error:

```
*** ERROR ***
The stateful material property 'elastic_strain' was declared in ComputeFiniteStrainElasticStress "elasticStrain" but was not declared in that object on checkpoint.

This is not currently supported due to ambiguity.

Please contact the development team on GitHub if you desire this capability.
```
I don't even understand what this capability is. Both the original file and the new restart file have a ComputeFiniteStrainElasticStress object.

**Reply by GiudGiud** [reply, marked as answer]
Does it have exactly the same name? And same suffix/prefix (if those are used) for material properties?

**Reply by selarem** [reply]
Now the error is : 
```
*** ERROR ***
The stateful material properties in ComputeLinearElasticPFFractureStress "elastic" that are being restarted do not match the stored properties in the same material object from the checkpoint.

Checkpointed stateful properties:
 - hist

Current stateful properties:
 - hist
 - stress
```
We use the same material in both computations.

**Reply by GiudGiud** [reply]
Some option must be different. 
Can you paste the Materials and GlobalParams blocks in both simulations?

**Reply by selarem** [reply]
Everything is the same only the Modules:
The first (to import) is quasistatic:  
```
[Physics/SolidMechanics/QuasiStatic]
  [all]
    strain = small 
    add_variables = true
    extra_vector_tags = 'kernel_residual'
    additional_generate_output = 'stress_xx stress_xy stress_yy'
  []
[]

```

and the main that imports : 


```
[Modules]
  [./TensorMechanics/DynamicMaster]
      [all]
        add_variables = true
        hht_alpha =-0.333
        newmark_beta = 0.44 #0.3025
        newmark_gamma =0.83
        strain = small 
        density = 1180 
        additional_generate_output = 'stress_yy stress_xy stress_xx'
       extra_vector_tags = 'kernel_residual'
      []
  [../]
[]
```

> Some option must be different. Can you paste the Materials and GlobalParams blocks in both simulations?

**Reply by GiudGiud** [reply]
the executioner is a Transient for both cases too?

**Reply by selarem** [reply]
> the executioner is a Transient for both cases too?

yes, for both :
```
[Executioner]
  type = Transient
solve_type = newton
petsc_options_iname = '-pc_type -pc_factor_mat_solver_package -snes_type'
  petsc_options_value = 'lu       superlu_dist                  vinewtonrsls'
```

**Reply by GiudGiud** [reply]
can you paste this object of type ComputeLinearElasticPFFractureStress?
the Materials block you posted was different

**Reply by selarem** [reply]
An aspect that could help in debugging is as follow:
With a material block
```
[Materials]
 [./elasticity_tensor]
    type = ComputeIsotropicElasticityTensor
    youngs_modulus = 3.09e9
    poissons_ratio = 0.35
  [../]
[./stress]
    type = ComputeFiniteStrainElasticStress
  [../]
[]
```
A static analysis with : 
```
 [Physics/SolidMechanics/QuasiStatic]
  [all]
    strain = FINITE
    add_variables = true
    extra_vector_tags = 'kernel_residual'
   additional_generate_output = 'stress_xx stress_xy stress_yy'
  []
[]
```
works fine. However, when  we change to  

```
[Modules]
  [./TensorMechanics/DynamicMaster]
      [all]
        add_variables = true
        hht_alpha =-0.333
        newmark_beta = 0.44 #0.3025
        newmark_gamma =0.83
        strain = FINITE 
         density = 1180 
        additional_generate_output = 'stress_yy stress_xy stress_xx'
       extra_vector_tags = 'kernel_residual'
      []
  [../]
[]
```
We obtain the Error : 
```

*** ERROR ***
The number of states for the restarted stateful material property 'stress' do not match.

Checkpointed states: 1
Current states: 2
```

**Reply by GiudGiud** [reply]
Ok this is yet another error now
With the material you initially mentioned 


But this one is fixable. In the static calculation, make sure there is a material that requests the older state of the properties that are stateful. 

Either find one that does or add it (just create an object that calls getMaterialPropertyOlder)

**Reply by selarem** [reply]
> Ok this is yet another error now With the material you initially mentioned
> 
> But this one is fixable. In the static calculation, make sure there is a material that requests the older state of the properties that are stateful.
> 
> Either find one that does or add it (just create an object that calls getMaterialPropertyOlder)

After verifications, it looks like we should use the same   analysis type   (either dynamic or static) for both computations whatever the material is.
But, we want a static analysis for the preload analysis, and dynamic for the main analysis (that imports the preload results).
Do you think that by creating a new material class  getMaterialPropertyOld, things could be fixed ? 
And in this case what to get as Property ?

**Reply by GiudGiud** [reply]
I think it will clear this error

The property to get (with older not old) should be the one you were getting an error for

**Reply by selarem** [reply]
The main problem remain not solved ! We just wanted to have IC on displacements dof in a dynamic analysis. 
Why this needs all these complications ! There should be an easier way to do it I think. 

> I think it will clear this error
> 
> The property to get (with older not old) should be the one you were getting an error for

**Reply by GiudGiud** [reply]
Did it clear the error?
What error are you getting now?

**Reply by selarem** [reply]
I was unable to implement it because I don't fully understand why such a mechanically complex solution is necessary. In a Cauchy problem, initial conditions (IC) can typically be specified in terms of displacements or velocities. In our case, we are looking for an initial condition on displacements that originates from another computation. Why is this so complex to implement?"

**Reply by GiudGiud** [reply]
It's not complicated you are just unwilling to work with the software. 
Hundreds of students have gotten their research done with this tool, the complexity is manageable

**Reply by selarem** [reply]
> It's not complicated you are just unwilling to work with the software. Hundreds of students have gotten their research done with this tool, the complexity is manageable

With this software I do not understand why, strain and  stress are zero while the IC condition is respected ? This works only in "small strain" but not with "Finite strain".  Before trying anything, we want to understand this first.

[Variables]
  [./disp_x]
    [../]
  [./disp_y]
  [./InitialCondition]
      type = FunctionIC
      function = parsed_function
    [../]
  [../]
[./c] 
  [../] 
[]

**Reply by GiudGiud** [reply]
Strain and stress are stateful material properties. They depend on the prior state (prior time step if you will), and using an exodus restart or an IC does not set those. 

You can use Checkpoint restart, or you can figure out the, currently complex, restart using projected stateful material properties
